Libraries/LibWeb/Fetch/Body.cpp

```diff
@@ -231,4 +237,227 @@ WebIDL::ExceptionOr<GC::Ref<WebIDL::Promise>> consume_body(JS::Realm& realm, Bod
+            header_value = header_value.trim(Infrastructure::HTTP_TAB_OR_SPACE, TrimMode::Right);
+
+            // 3. Set contentType to the isomorphic decoding of header value.
+            header.content_type = Infra::isomorphic_decode(header_value.bytes());
+        }
+        // -> Otherwise
+        else {
+            // 1. Collect a sequence of bytes that are not 0x0A (LF) or 0x0D (CR), given position. (Do nothing with those bytes.)
+            lexer.ignore_until(Infrastructure::is_http_newline);
+        }
+
+        // 9. If position does not point to a sequence of bytes starting with 0x0D 0x0A (CR LF), return failure. Otherwise, advance position by 2 (past the newline).
+        if (!lexer.consume_specific("\r\n"sv))
+            return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, "Expected CRLF 3"sv };
+    }
+    return header;
+}
+
+// https://andreubotella.github.io/multipart-form-data/#multipart-form-data-parser
+WebIDL::ExceptionOr<Vector<XHR::FormDataEntry>> parse_multipart_form_data(JS::Realm& realm, StringView input, MimeSniff::MimeType const& mime_type)
+{
+    // 1. Assert: mimeType’s essence is "multipart/form-data".
+    VERIFY(mime_type.essence() == "multipart/form-data"sv);
+
+    // 2. If mimeType’s parameters["boundary"] does not exist, return failure. Otherwise, let boundary be the result of UTF-8 decoding mimeType’s parameters["boundary"].
+    auto maybe_boundary = mime_type.parameters().get("boundary"sv);
+    if (!maybe_boundary.has_value())
+        return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, "Mime type must have a 'boundary' parameter"sv };
+    auto boundary = maybe_boundary.release_value();
+
+    // 3. Let entry list be an empty entry list.
+    Vector<XHR::FormDataEntry> entry_list;
+
+    // 4. Let position be a pointer to a byte in input, initially pointing at the first byte.
+    GenericLexer lexer(input);
+
+    auto boundary_with_dashes = MUST(String::formatted("--{}", boundary));
+
+    // 5. While true:
+    while (true) {
+        // 1. If position points to a sequence of bytes starting with 0x2D 0x2D (`--`) followed by boundary, advance position by 2 + the length of boundary. Otherwise, return failure.
+        if (!lexer.consume_specific(boundary_with_dashes))
+            return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, "Boundary not found"sv };
+
+        // 2. If position points to the sequence of bytes 0x2D 0x2D 0x0D 0x0A (`--` followed by CR LF) followed by the end of input, return entry list.
+        if (lexer.next_is("--\r\n"sv))
+            return entry_list;
+
+        // 3. If position does not point to a sequence of bytes starting with 0x0D 0x0A (CR LF), return failure.
+        // 4. Advance position by 2. (This skips past the newline.)
+        if (!lexer.consume_specific("\r\n"sv))
+            return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, "Expected CRLF 1"sv };
```

Comment:
maybe it is possible to provide more context in the error message instead of a number (same below)?